File: src/lib/qr-parser.ts
export function parseQRCodeData(data: string) {
  let type: string = "text";
  let content = data;
  let details: any = {};

  if (/^(http|https):\/\/[^ "]+$/.test(data)) {
    type = "url";
    if (data.includes("instagram.com") || data.includes("forms.gle")) {
      type = "social";
    }
  } else if (data.startsWith("WIFI:")) {
    type = "wifi";
    const sMatch = data.match(/S:([^;]+)/);
    const pMatch = data.match(/P:([^;]+)/);
    details = {
      ssid: sMatch ? sMatch[1] : "Unknown",
      password: pMatch ? pMatch[1] : ""
    };
  } else if (data.toUpperCase().startsWith("BEGIN:VCARD")) {
    type = "vcard";
    const nameMatch = data.match(/FN:(.+)/i);
    const orgMatch = data.match(/ORG:(.+)/i);
    const telMatch = data.match(/TEL.*:(.+)/i);
    details = {
      name: nameMatch ? nameMatch[1].trim() : "Contact",
      org: orgMatch ? orgMatch[1].trim() : "Unknown Org",
      tel: telMatch ? telMatch[1].trim() : ""
    };
  } else if (data.toUpperCase().startsWith("BEGIN:VEVENT")) {
    type = "event";
    const summaryMatch = data.match(/SUMMARY:(.+)/i);
    details = {
       summary: summaryMatch ? summaryMatch[1].trim() : "Event"
    };
  } else if (data.toLowerCase().startsWith("geo:")) {
    type = "geo";
    const parts = data.replace("geo:", "").split(",");
    details = {
      lat: parts[0],
      lng: parts[1]
    }
  } else if (/^(978|979)\d{10}$/.test(data)) {
    type = "isbn";
  } else if (/^\d{12}$/.test(data)) {
    type = "upc";
  } else if (/^\d{13}$/.test(data)) {
    type = "barcode"; // EAN-13
  } else if (/^[\x00-\x7F]+$/.test(data) && data.length > 5 && data.length < 40 && !data.includes(" ")) {
    type = "barcode"; // Generic CODE-128
  }

  return { type, content, details };
}

File: app/src/main/java/com/example/qrmax/data/parser/QRParser.kt
package com.example.qrmax.data.parser

object QRParser {
    fun parse(data: String): ParsedResult {
        val type: String
        val details = mutableMapOf<String, String>()

        when {
            Regex("^(http|https)://[^ ]+$").matches(data) -> {
                type = if (data.contains("instagram.com") || data.contains("forms.gle")) "social" else "url"
            }
            data.startsWith("WIFI:", ignoreCase = true) -> {
                type = "wifi"
                val sMatch = Regex("S:([^;]+)").find(data)
                val pMatch = Regex("P:([^;]+)").find(data)
                details["ssid"] = sMatch?.groupValues?.get(1) ?: "Unknown"
                details["password"] = pMatch?.groupValues?.get(1) ?: ""
            }
            data.uppercase().startsWith("BEGIN:VCARD") -> {
                type = "vcard"
                val nameMatch = Regex("FN:(.+)", RegexOption.IGNORE_CASE).find(data)
                val orgMatch = Regex("ORG:(.+)", RegexOption.IGNORE_CASE).find(data)
                val telMatch = Regex("TEL.*:(.+)", RegexOption.IGNORE_CASE).find(data)
                details["name"] = nameMatch?.groupValues?.get(1)?.trim() ?: "Contact"
                details["org"] = orgMatch?.groupValues?.get(1)?.trim() ?: "Unknown Org"
                details["tel"] = telMatch?.groupValues?.get(1)?.trim() ?: ""
            }
            data.uppercase().startsWith("BEGIN:VEVENT") -> {
                type = "event"
                val summaryMatch = Regex("SUMMARY:(.+)", RegexOption.IGNORE_CASE).find(data)
                details["summary"] = summaryMatch?.groupValues?.get(1)?.trim() ?: "Event"
            }
            data.lowercase().startsWith("geo:") -> {
                type = "geo"
                val parts = data.removePrefix("geo:").split(",")
                if (parts.size >= 2) {
                    details["lat"] = parts[0]
                    details["lng"] = parts[1]
                }
            }
            Regex("^(978|979)\\d{10}$").matches(data) -> type = "isbn"
            Regex("^\\d{12}$").matches(data) -> type = "upc"
            Regex("^\\d{13}$").matches(data) -> type = "barcode"
            Regex("^[\\x00-\\x7F]+$").matches(data) && data.length in 6..39 && !data.contains(" ") -> type = "barcode"
            else -> type = "text"
        }

        return ParsedResult(type = type, content = data, details = details)
    }
}

data class ParsedResult(
    val type: String,
    val content: String,
    val details: Map<String, String> = emptyMap()
)
